internal/server/v2/observation/calculation.go

```diff
@@ -17,56 +17,137 @@ package observation
 
 import (
 	"context"
+	"net/http"
 
+	pb "github.com/datacommonsorg/mixer/internal/proto"
 	pbv2 "github.com/datacommonsorg/mixer/internal/proto/v2"
 	"github.com/datacommonsorg/mixer/internal/server/cache"
+	"github.com/datacommonsorg/mixer/internal/server/resource"
 	"github.com/datacommonsorg/mixer/internal/store"
 )
 
-// CalculateObservationResponses returns a list of ObservationResponses based
-// on StatisticalCalculation formulas for the missing data in an input
-// ObservationResponse.
-func CalculateObservationResponses(
+// Computes a calculation for a variable and entity, based on a formula and input data.
+func Calculate(
 	ctx context.Context,
 	store *store.Store,
-	inputResp *pbv2.ObservationResponse,
 	cachedata *cache.Cache,
-) []*pbv2.ObservationResponse {
-	calculatedResponses := []*pbv2.ObservationResponse{}
-	for variable, variableObservation := range inputResp.ByVariable {
+	metadata *resource.Metadata,
+	httpClient *http.Client,
+	formula string,
+	variable string,
+	entity *pbv2.DcidOrExpression,
+	inputReq *pbv2.ObservationRequest,
+) (*pbv2.ObservationResponse, error) {
+	variableFormula, err := NewVariableFormula(formula)
+	if err != nil {
+		return nil, err
+	}
+	newReq := &pbv2.ObservationRequest{
+		Variable: &pbv2.DcidOrExpression{Dcids: variableFormula.StatVars},
+		Entity:   entity,
+		Date:     inputReq.GetDate(),
+		Value:    inputReq.GetValue(),
+		Filter:   inputReq.GetFilter(),
+		Select:   inputReq.GetSelect(),
+	}
+	variableResp, err := ObservationInternal(ctx, store, cachedata, metadata, httpClient, newReq)
+	if err != nil {
+		return nil, err
+	}
+	computeLeafObs(variableResp, variableFormula)
```

Comment:
now the VariableFormula is not a formula anymore, but the actual tree with observation.

I wonder if it's possible to move `CandidateObs` out of `ASTNode`?

Reply:
Done